File: Simple1ExoPlayer/app/src/main/java/com/daeseong/simple1exoplayer/MainActivity.java
package com.daeseong.simple1exoplayer;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import android.Manifest;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.ServiceConnection;
import android.content.pm.ActivityInfo;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.media.session.PlaybackState;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.IBinder;
import android.telephony.TelephonyManager;
import android.util.Log;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.ImageButton;
import android.widget.SeekBar;
import android.widget.TextView;
import java.util.ArrayList;
import java.util.Formatter;
import java.util.Locale;
import java.util.Timer;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = MainActivity.class.getSimpleName();

    private ImageButton btnPre, btnPlay, btnPause, btnNext, btnPrevious, btnNextgo, btnSearch;
    private TextView txtStartTime, txtEndTime, txtDesc;
    private SeekBar seekBar;
    private PlayerTimer playerTimer;

    private MusicService musicService;
    private BroadcastReceiver broadcastReceiver;
    private Intent intentservice;

    private ArrayList<MusicInfo> musicList  = new ArrayList<MusicInfo>();
    private int CurrentPlayIndex = -1;

    private MarqueeTask taskMarquee;
    private Timer timerMarquee = null;

    public ActivityResultLauncher<String[]> requestPermissions;

    private static final String[] PERMISSIONS = new String[]{Manifest.permission.READ_PHONE_STATE, Manifest.permission.READ_EXTERNAL_STORAGE};
    private static final String[] PERMISSIONS33 = new String[]{Manifest.permission.READ_PHONE_STATE, Manifest.permission.READ_MEDIA_AUDIO};

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        InitTitleBar();

        setContentView(R.layout.activity_main);

        initPermissionsLauncher();

        txtStartTime = findViewById(R.id.startTime);
        txtEndTime = findViewById(R.id.endTime);
        txtDesc = findViewById(R.id.tvDesc);

        btnSearch = findViewById(R.id.btnSearch);
        btnSearch.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                checkPermissions();

                //음악 폴더 선택
                getMusicList item = new getMusicList();
                musicList = item.getData();

                if (musicList.size() == 0) return;

                CurrentPlayIndex = 0;

                //Marquee
                txtDesc.setText(musicList.get(CurrentPlayIndex).getMusicName());

                Uri uri = Uri.parse(musicList.get(CurrentPlayIndex).getMusicPath());
                playPlayer(uri);
                btnPlay.setVisibility(View.INVISIBLE);
                btnPause.setVisibility(View.VISIBLE);
            }
        });

        //이전곡
        btnPre = findViewById(R.id.btnPre);
        btnPre.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                if(musicList.size() == 0) return;

                CurrentPlayIndex--;

                if (CurrentPlayIndex < 0)
                    CurrentPlayIndex = musicList.size() - 1;

                //Marquee
                txtDesc.setText(musicList.get(CurrentPlayIndex).getMusicName());

                Uri uri = Uri.parse(musicList.get(CurrentPlayIndex).getMusicPath());
                playPlayer(uri);
                btnPlay.setVisibility(View.INVISIBLE);
                btnPause.setVisibility(View.VISIBLE);
            }
        });

        //다음곡
        btnNext = findViewById(R.id.btnNext);
        btnNext.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                if(musicList.size() == 0) return;

                CurrentPlayIndex++;

                if (CurrentPlayIndex > (musicList.size() - 1))
                    CurrentPlayIndex = 0;

                //Marquee
                txtDesc.setText(musicList.get(CurrentPlayIndex).getMusicName());

                Uri uri = Uri.parse(musicList.get(CurrentPlayIndex).getMusicPath());
                playPlayer(uri);
                btnPlay.setVisibility(View.INVISIBLE);
                btnPause.setVisibility(View.VISIBLE);
            }
        });

        //연주
        btnPlay = findViewById(R.id.btnPlay);
        btnPlay.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                setPlayWhenReady(true);
                btnPlay.setVisibility(View.INVISIBLE);
                btnPause.setVisibility(View.VISIBLE);
            }
        });

        //일시정지
        btnPause = findViewById(R.id.btnPause);
        btnPause.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                setPlayWhenReady(false);
                btnPlay.setVisibility(View.VISIBLE);
                btnPause.setVisibility(View.INVISIBLE);
            }
        });

        //진행바
        seekBar = (SeekBar)findViewById(R.id.seekBar);
        seekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {

                if(!fromUser) return;

                seekTo(progress * 1000);
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
                setPlayWhenReady(false);
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                setPlayWhenReady(true);
            }
        });

        checkPermissions();
    }

    private ServiceConnection serviceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder iBinder) {
            MusicService.MusicBinder mServiceBinder = (MusicService.MusicBinder) iBinder;
            musicService = mServiceBinder.getMusicService();
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            System.exit(0);
        }
    };

    private void InitTitleBar(){

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            Window window = getWindow();
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
            window.setStatusBarColor(Color.rgb(255, 255, 255));
        }

        try {
            //안드로이드 8.0 오레오 버전에서만 오류 발생
            setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
            supportRequestWindowFeature(Window.FEATURE_NO_TITLE);
        } catch (Exception ex) {
            Log.e(TAG, ex.getMessage().toString());
        }
    }

    private void checkPermissions() {

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            boolean bPermissResult = false;

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {

                for (String permission : PERMISSIONS33) {
                    bPermissResult = checkSelfPermission(permission) == PackageManager.PERMISSION_GRANTED;
                    if (!bPermissResult) {
                        break;
                    }
                }

                if (!bPermissResult) {
                    requestPermissions.launch(PERMISSIONS33);
                } else {
                    Log.e(TAG, "PERMISSIONS33 권한 소유-SimpleExoPlayer 초기화");
                    initalizePlayer();
                }

            } else {

                for (String permission : PERMISSIONS) {
                    bPermissResult = checkSelfPermission(permission) == PackageManager.PERMISSION_GRANTED;
                    if (!bPermissResult) {
                        break;
                    }
                }

                if (!bPermissResult) {
                    requestPermissions.launch(PERMISSIONS);
                } else {
                    Log.e(TAG, "PERMISSIONS 권한 소유-SimpleExoPlayer 초기화");
                    initalizePlayer();
                }
            }
        }
    }

    private void initPermissionsLauncher() {

        requestPermissions = registerForActivityResult(new ActivityResultContracts.RequestMultiplePermissions(), result -> {

            boolean bPhone = false;
            boolean bAudio = false;

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {

                bPhone = Boolean.TRUE.equals(result.get(Manifest.permission.READ_PHONE_STATE));
                bAudio = Boolean.TRUE.equals(result.get(Manifest.permission.READ_MEDIA_AUDIO));

            } else {

                bPhone = Boolean.TRUE.equals(result.get(Manifest.permission.READ_PHONE_STATE));
                bAudio = Boolean.TRUE.equals(result.get(Manifest.permission.READ_EXTERNAL_STORAGE));
            }

            if (bPhone && bAudio) {
                Log.e(TAG, "PERMISSIONS 권한 소유-SimpleExoPlayer 초기화");
                initalizePlayer();
            } else {
                Log.e(TAG, "PERMISSIONS 권한 미소유");
            }
        });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        musicList.clear();
        closeMarqueeTimer();
        stopplayerTimer();
        releasePlayer();

        if (broadcastReceiver != null) {
            unregisterReceiver(broadcastReceiver);
        }

        stopService(intentservice);
        unbindService(serviceConnection);
    }

    private void initalizePlayer() {

        broadcastReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {

                TelephonyManager tm = (TelephonyManager) context.getSystemService(Service.TELEPHONY_SERVICE);
                if (tm != null) {
                    if (tm.getCallState() == TelephonyManager.CALL_STATE_RINGING) {

                        if(isPlaying()) {
                            setPlayWhenReady(false);
                            btnPlay.setVisibility(View.VISIBLE);
                            btnPause.setVisibility(View.INVISIBLE);
                        }

                    }
                }

                int playerState = intent.getIntExtra("state", 0);
                if (playerState == PlaybackState.STATE_BUFFERING) {
                    Log.e(TAG, "PlaybackState.STATE_BUFFERING");
                } else if (playerState == PlaybackState.STATE_PLAYING) {
                    Log.e(TAG, "PlaybackState.STATE_PLAYING");
                } else if (playerState == PlaybackState.STATE_PAUSED) {
                    Log.e(TAG, "PlaybackState.STATE_PAUSED");
                } else if (playerState == PlaybackState.STATE_NONE) {

                    if(!isPlaying()){

                        if(musicList.size() == 0) return;

                        CurrentPlayIndex++;

                        if (CurrentPlayIndex > (musicList.size() - 1))
                            CurrentPlayIndex = 0;

                        //Marquee
                        txtDesc.setText(musicList.get(CurrentPlayIndex).getMusicName());

                        Uri uri = Uri.parse(musicList.get(CurrentPlayIndex).getMusicPath());
                        playPlayer(uri);
                        btnPlay.setVisibility(View.INVISIBLE);
                        btnPause.setVisibility(View.VISIBLE);
                    }
                }

            }
        };
        IntentFilter filter = new IntentFilter();
        filter.addAction("android.intent.action.PHONE_STATE");
        filter.addAction("android.net.conn.CONNECTIVITY_CHANGE");
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            registerReceiver(broadcastReceiver, filter, Context.RECEIVER_NOT_EXPORTED);
        } else {
            registerReceiver(broadcastReceiver, filter);
        }

        //music service
        intentservice = new Intent(MainActivity.this, MusicService.class);
        bindService(intentservice, serviceConnection, Context.BIND_AUTO_CREATE);
        LocalBroadcastManager.getInstance(this).registerReceiver((broadcastReceiver),  new IntentFilter("com.daeseong.simple1exoplayer.PLAYER_STATUS"));
    }

    private void releasePlayer(){
        if(musicService != null ){
            musicService.releasePlayer();
        }
    }

    private void playPlayer(String sUrl){
        if(musicService != null ){
            musicService.playPlayer(sUrl);
            setSeekBarProgress();
        }
    }

    private void playPlayer(Uri uri){
        if(musicService != null ){
            musicService.playPlayer(uri);
            setSeekBarProgress();
        }
    }

    private void stopPlayer(){
        if(musicService != null ){
            musicService.stopPlayer();
        }
    }

    private boolean isPlaying(){
        if(musicService != null ){
            return  musicService.isPlaying();
        }
        return false;
    }

    private void PreplayPlayer(){
        if(musicService != null ){
            musicService.prePlayPlayer();
        }
    }

    private void NextplayPlayer(){
        if(musicService != null ){
            musicService.nextPlayPlayer();
        }
    }

    private void setPlayWhenReady(boolean bReady){
        if(musicService != null ){
            musicService.setPlayWhenReady(bReady);
        }
    }

    private long getCurrentPosition(){
        long position = 0;
        if(musicService != null ){
            position = musicService.getCurrentPosition();
        }
        return position;
    }

    private long getDuration(){
        long duration = 0;
        if(musicService != null ){
            duration = musicService.getDuration();
        }
        return duration;
    }

    private void seekTo(long progress){
        if(musicService != null ){
            musicService.seekTo(progress);
        }
    }

    private String stringForTime(int timeMs) {
        StringBuilder mFormatBuilder;
        Formatter mFormatter;
        mFormatBuilder = new StringBuilder();
        mFormatter = new Formatter(mFormatBuilder, Locale.getDefault());
        int totalSeconds =  timeMs / 1000;

        int seconds = totalSeconds % 60;
        int minutes = (totalSeconds / 60) % 60;
        int hours   = totalSeconds / 3600;

        mFormatBuilder.setLength(0);
        if (hours > 0) {
            return mFormatter.format("%d:%02d:%02d", hours, minutes, seconds).toString();
        } else {
            return mFormatter.format("%02d:%02d", minutes, seconds).toString();
        }
    }

    private void stopplayerTimer(){
        if (playerTimer != null) {
            playerTimer.stop();
            playerTimer = null;
        }
    }

    private void setSeekBarProgress(){

        //Marquee
        startMarqueeTimer();

        stopplayerTimer();

        playerTimer = new PlayerTimer();
        playerTimer.setCallback(new PlayerTimer.Callback() {
            @Override
            public void onTick(long timeMillis) {

                if (isPlaying()) {

                    long position = getCurrentPosition();
                    long duration = getDuration();

                    if (duration <= 0) return;

                    seekBar.setMax((int) duration / 1000);
                    seekBar.setProgress((int) position / 1000);

                    txtStartTime.setText(stringForTime((int) getCurrentPosition()));
                    txtEndTime.setText(stringForTime((int) getDuration()));
                }
            }
        });
        playerTimer.start();
    }

    private void closeMarqueeTimer(){
        if (timerMarquee != null) {
            timerMarquee.cancel();
            timerMarquee = null;
        }
    }

    private void startMarqueeTimer(){
        closeMarqueeTimer();
        taskMarquee = new MarqueeTask(txtDesc);
        timerMarquee = new Timer();
        timerMarquee.schedule(taskMarquee, 0, 10000);
    }
}

File: Simple1ExoPlayer(kotlin)/app/src/main/java/com/daeseong/simple1exoplayer/MainActivity.kt
package com.daeseong.simple1exoplayer

import android.Manifest.permission.*
import android.content.*
import android.content.pm.ActivityInfo
import android.content.pm.PackageManager
import android.graphics.Color
import android.media.session.PlaybackState
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.IBinder
import android.telephony.TelephonyManager
import android.util.Log
import android.view.View
import android.view.Window
import android.view.WindowManager
import android.widget.ImageButton
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.daeseong.simple1exoplayer.MusicService.MusicBinder
import java.util.*
import kotlin.system.exitProcess

class MainActivity : AppCompatActivity() {

    private val tag = MainActivity::class.java.simpleName

    private lateinit var btnPre: ImageButton
    private lateinit var btnPlay: ImageButton
    private lateinit var btnPause: ImageButton
    private lateinit var btnNext: ImageButton
    private lateinit var btnSearch: ImageButton
    private lateinit var txtStartTime: TextView
    private lateinit var txtEndTime: TextView
    private lateinit var txtDesc: TextView
    private lateinit var seekBar: SeekBar

    private var playerTimer: PlayerTimer? = null
    private var musicService: MusicService? = null
    private var broadcastReceiver: BroadcastReceiver? = null
    private var intentservice: Intent? = null
    private var musicList = ArrayList<MusicInfo>()
    private var CurrentPlayIndex = -1
    private var taskMarquee: MarqueeTask? = null
    private var timerMarquee: Timer? = null

    private lateinit var requestPermissions: ActivityResultLauncher<Array<String>>

    private val PERMISSIONS = arrayOf(READ_PHONE_STATE, READ_EXTERNAL_STORAGE)
    private val PERMISSIONS33 = arrayOf(READ_PHONE_STATE, READ_MEDIA_AUDIO)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        InitTitleBar()

        setContentView(R.layout.activity_main)

        initPermissionsLauncher()

        txtStartTime = findViewById(R.id.startTime)
        txtEndTime = findViewById(R.id.endTime)
        txtDesc = findViewById(R.id.tvDesc)
        btnSearch = findViewById(R.id.btnSearch)

        btnSearch.setOnClickListener(View.OnClickListener {

            checkPermissions()

            //음악 폴더 선택
            val item = getMusicList()
            musicList = item.getData()
            if (musicList.size == 0) return@OnClickListener
            CurrentPlayIndex = 0

            //Marquee
            txtDesc.text = musicList[CurrentPlayIndex].musicName
            val uri = Uri.parse(musicList[CurrentPlayIndex].musicPath)
            playPlayer(uri)
            btnPlay.visibility = View.INVISIBLE
            btnPause.visibility = View.VISIBLE
        })

        //이전곡
        btnPre = findViewById(R.id.btnPre)
        btnPre.setOnClickListener(View.OnClickListener {

            if (musicList.size == 0) return@OnClickListener
            CurrentPlayIndex--
            if (CurrentPlayIndex < 0) CurrentPlayIndex = musicList.size - 1

            //Marquee
            txtDesc.text = musicList[CurrentPlayIndex].musicName
            val uri = Uri.parse(musicList[CurrentPlayIndex].musicPath)
            playPlayer(uri)
            btnPlay.visibility = View.INVISIBLE
            btnPause.visibility = View.VISIBLE
        })

        //다음곡
        btnNext = findViewById(R.id.btnNext)
        btnNext.setOnClickListener(View.OnClickListener {

            if (musicList.size == 0) return@OnClickListener
            CurrentPlayIndex++
            if (CurrentPlayIndex > musicList.size - 1) CurrentPlayIndex = 0

            //Marquee
            txtDesc.text = musicList[CurrentPlayIndex].musicName
            val uri = Uri.parse(musicList[CurrentPlayIndex].musicPath)
            playPlayer(uri)
            btnPlay.visibility = View.INVISIBLE
            btnPause.visibility = View.VISIBLE
        })

        //연주
        btnPlay = findViewById(R.id.btnPlay)
        btnPlay.setOnClickListener(View.OnClickListener {

            setPlayWhenReady(true)
            btnPlay.visibility = View.INVISIBLE
            btnPause.visibility = View.VISIBLE
        })

        //일시정지
        btnPause = findViewById(R.id.btnPause)
        btnPause.setOnClickListener(View.OnClickListener {

            setPlayWhenReady(false)
            btnPlay.visibility = View.VISIBLE
            btnPause.visibility = View.INVISIBLE
        })

        //진행바
        seekBar = findViewById<View>(R.id.seekBar) as SeekBar
        seekBar.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                if (!fromUser) return
                seekTo((progress * 1000).toLong())
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {
                setPlayWhenReady(false)
            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {
                setPlayWhenReady(true)
            }
        })
        checkPermissions()
    }

    private val serviceConnection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, iBinder: IBinder) {
            val mServiceBinder = iBinder as MusicBinder
            musicService = mServiceBinder.getMusicService()
        }

        override fun onServiceDisconnected(name: ComponentName) {
            exitProcess(0)
        }
    }

    private fun InitTitleBar() {

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            val window = window
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
            window.statusBarColor = Color.rgb(255, 255, 255)
        }

        try {
            //안드로이드 8.0 오레오 버전에서만 오류 발생
            requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
            supportRequestWindowFeature(Window.FEATURE_NO_TITLE)
        } catch (ex: Exception) {
            Log.e(tag, ex.message.toString())
        }
    }

    private fun checkPermissions() {

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {

            var bPermissResult = false

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                for (permission in PERMISSIONS33) {
                    bPermissResult = checkSelfPermission(permission) == PackageManager.PERMISSION_GRANTED
                    if (!bPermissResult) {
                        break
                    }
                }

                if (!bPermissResult) {
                    requestPermissions.launch(PERMISSIONS33)
                } else {
                    Log.e(tag, "PERMISSIONS33 권한 소유-SimpleExoPlayer 초기화")
                    initalizePlayer()
                }
            } else {
                for (permission in PERMISSIONS) {
                    bPermissResult = checkSelfPermission(permission) == PackageManager.PERMISSION_GRANTED
                    if (!bPermissResult) {
                        break
                    }
                }
                if (!bPermissResult) {
                    requestPermissions.launch(PERMISSIONS)
                } else {
                    Log.e(tag, "PERMISSIONS 권한 소유-SimpleExoPlayer 초기화")
                    initalizePlayer()
                }
            }
        }
    }

    private fun initPermissionsLauncher() {
        requestPermissions = registerForActivityResult(ActivityResultContracts.RequestMultiplePermissions()) { result ->

            var bPhone = false
            var bAudio = false

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                bPhone =  java.lang.Boolean.TRUE == result[READ_PHONE_STATE]
                bAudio =  java.lang.Boolean.TRUE == result[READ_MEDIA_AUDIO]
            } else {
                bPhone =  java.lang.Boolean.TRUE == result[READ_PHONE_STATE]
                bAudio =  java.lang.Boolean.TRUE == result[READ_EXTERNAL_STORAGE]
            }

            if (bPhone && bAudio) {
                Log.e(tag,"PERMISSIONS 권한 소유-SimpleExoPlayer 초기화")
                initalizePlayer()
            } else {
                Log.e(tag, "PERMISSIONS 권한 미소유")
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()

        musicList.clear()
        closeMarqueeTimer()
        stopplayerTimer()
        releasePlayer()

        if (broadcastReceiver != null) {
            unregisterReceiver(broadcastReceiver)
        }

        stopService(intentservice)
        unbindService(serviceConnection)
    }

    private fun initalizePlayer() {

        broadcastReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {

                val tm = context.getSystemService(TELEPHONY_SERVICE) as TelephonyManager
                if (tm != null) {

                    if (tm.callState == TelephonyManager.CALL_STATE_RINGING) {

                        if (isPlaying) {
                            setPlayWhenReady(false)
                            btnPlay.visibility = View.VISIBLE
                            btnPause.visibility = View.INVISIBLE
                        }
                    }
                }

                val playerState = intent.getIntExtra("state", 0)
                if (playerState == PlaybackState.STATE_BUFFERING) {

                    Log.e(tag, "PlaybackState.STATE_BUFFERING")
                } else if (playerState == PlaybackState.STATE_PLAYING) {

                    Log.e(tag, "PlaybackState.STATE_PLAYING")
                } else if (playerState == PlaybackState.STATE_PAUSED) {

                    Log.e(tag, "PlaybackState.STATE_PAUSED")
                } else if (playerState == PlaybackState.STATE_NONE) {

                    if (!isPlaying) {

                        if (musicList.size == 0) return
                        CurrentPlayIndex++
                        if (CurrentPlayIndex > musicList.size - 1) CurrentPlayIndex = 0

                        //Marquee
                        txtDesc!!.text = musicList[CurrentPlayIndex].musicName
                        val uri = Uri.parse(musicList[CurrentPlayIndex].musicPath)
                        playPlayer(uri)
                        btnPlay.visibility = View.INVISIBLE
                        btnPause.visibility = View.VISIBLE
                    }
                }
            }
        }

        val filter = IntentFilter()
        filter.addAction("android.intent.action.PHONE_STATE")
        filter.addAction("android.net.conn.CONNECTIVITY_CHANGE")
        registerReceiver(broadcastReceiver, filter)

        //music service
        intentservice = Intent(this@MainActivity, MusicService::class.java)
        bindService(intentservice, serviceConnection, BIND_AUTO_CREATE)
        LocalBroadcastManager.getInstance(this).registerReceiver(broadcastReceiver!!, IntentFilter("com.daeseong.simple1exoplayer.PLAYER_STATUS"))
    }

    private fun releasePlayer() {
        if (musicService != null) {
            musicService!!.releasePlayer()
        }
    }

    private fun playPlayer(sUrl: String) {
        if (musicService != null) {
            musicService!!.playPlayer(sUrl)
            setSeekBarProgress()
        }
    }

    private fun playPlayer(uri: Uri) {
        if (musicService != null) {
            musicService!!.playPlayer(uri)
            setSeekBarProgress()
        }
    }

    private fun stopPlayer() {
        if (musicService != null) {
            musicService!!.stopPlayer()
        }
    }

    private val isPlaying: Boolean
        private get() = if (musicService != null) {
            musicService!!.isPlaying()
        } else false

    private fun PreplayPlayer() {
        if (musicService != null) {
            musicService!!.preplayPlayer()
        }
    }

    private fun NextplayPlayer() {
        if (musicService != null) {
            musicService!!.nextplayPlayer()
        }
    }

    private fun setPlayWhenReady(bReady: Boolean) {
        if (musicService != null) {
            musicService!!.setPlayWhenReady(bReady)
        }
    }

    private val currentPosition: Long
        private get() {
            var position: Long = 0
            if (musicService != null) {
                position = musicService!!.getCurrentPosition()
            }
            return position
        }

    private val duration: Long
        private get() {
            var duration: Long = 0
            if (musicService != null) {
                duration = musicService!!.getDuration()
            }
            return duration
        }

    private fun seekTo(progress: Long) {
        if (musicService != null) {
            musicService!!.seekTo(progress)
        }
    }

    private fun stringForTime(timeMs: Int): String {
        val mFormatter: Formatter
        val mFormatBuilder: StringBuilder = StringBuilder()
        mFormatter = Formatter(mFormatBuilder, Locale.getDefault())
        val totalSeconds = timeMs / 1000
        val seconds = totalSeconds % 60
        val minutes = totalSeconds / 60 % 60
        val hours = totalSeconds / 3600
        mFormatBuilder.setLength(0)
        return if (hours > 0) {
            mFormatter.format("%d:%02d:%02d", hours, minutes, seconds).toString()
        } else {
            mFormatter.format("%02d:%02d", minutes, seconds).toString()
        }
    }

    private fun stopplayerTimer() {
        if (playerTimer != null) {
            playerTimer!!.stop()
            playerTimer = null
        }
    }

    private fun setSeekBarProgress() {

        //Marquee
        startMarqueeTimer()
        stopplayerTimer()
        playerTimer = PlayerTimer()
        playerTimer!!.setCallback(object : PlayerTimer.Callback {
            override fun onTick(timeMillis: Long) {

                if (isPlaying) {
                    val position = currentPosition
                    val duration = duration
                    if (duration <= 0) return
                    seekBar.max = duration.toInt() / 1000
                    seekBar.progress = position.toInt() / 1000
                    txtStartTime.text = stringForTime(currentPosition.toInt())
                    txtEndTime.text = stringForTime(duration.toInt())
                }
            }
        })
        playerTimer!!.start()
    }

    private fun closeMarqueeTimer() {
        if (timerMarquee != null) {
            timerMarquee!!.cancel()
            timerMarquee = null
        }
    }

    private fun startMarqueeTimer() {
        closeMarqueeTimer()
        taskMarquee = MarqueeTask(txtDesc)
        timerMarquee = Timer()
        timerMarquee!!.schedule(taskMarquee, 0, 10000)
    }
}